libsignal-service/src/sealed_session_cipher.rs

```diff
@@ -0,0 +1,1070 @@
+    key_id: u32,
+    key: PublicKey,
+    certificate: Vec<u8>,
+    signature: Vec<u8>,
+}
+
+#[derive(Debug, Clone)]
+struct EphemeralKeys {
+    chain_key: Vec<u8>,
+    cipher_key: Vec<u8>,
+    mac_key: Vec<u8>,
+}
+
+#[derive(Debug, Clone)]
+struct StaticKeys {
+    cipher_key: Vec<u8>,
+    mac_key: Vec<u8>,
+}
+
+#[derive(Debug, Clone)]
+pub struct CertificateValidator {
+    trust_root: PublicKey,
+}
+
+#[derive(Default, Debug, Clone)]
+pub(crate) struct DecryptionResult {
+    pub sender_uuid: Option<String>,
+    pub sender_e164: Option<String>,
+    pub device_id: i32,
+    pub padded_message: Vec<u8>,
+    pub version: u32,
+}
+
+impl UnidentifiedSenderMessage {
+    const CIPHERTEXT_VERSION: u8 = 1;
+
+    fn from_bytes(
+        context: &Context,
+        serialized: &[u8],
+    ) -> Result<Self, SealedSessionError> {
+        let version = (serialized[0] & 0xFF) >> 4;
+        if version > Self::CIPHERTEXT_VERSION {
+            return Err(SealedSessionError::InvalidMetadataVersionError(
+                version,
+            ));
+        }
+
+        let unidentified_sender_message: crate::proto::UnidentifiedSenderMessage =
+            prost::Message::decode(&serialized[1..serialized.len()])?;
+
+        match (
+            unidentified_sender_message.ephemeral_public,
+            unidentified_sender_message.encrypted_static,
+            unidentified_sender_message.encrypted_message,
+        ) {
+            (
+                Some(ephemeral_public),
+                Some(encrypted_static),
+                Some(encrypted_message),
+            ) => Ok(Self {
+                ephemeral: PublicKey::decode_point(
+                    &context,
+                    &ephemeral_public,
+                )?,
+                encrypted_static,
+                encrypted_message,
+            }),
+            _ => Err(SealedSessionError::InvalidMetadataMessageError(
+                "Missing fields".into(),
+            )),
+        }
+    }
+
+    fn into_bytes(self) -> Result<Vec<u8>, SealedSessionError> {
+        use prost::Message;
+        let mut buf = vec![];
+        buf.push(
+            (Self::CIPHERTEXT_VERSION << 4 | Self::CIPHERTEXT_VERSION) & 0xFF,
+        );
+        crate::proto::UnidentifiedSenderMessage {
+            ephemeral_public: Some(
+                self.ephemeral.to_bytes()?.as_slice().to_vec(),
+            ),
+            encrypted_static: Some(self.encrypted_static),
+            encrypted_message: Some(self.encrypted_message),
+        }
+        .encode(&mut buf)?;
+        Ok(buf)
+    }
+}
+
+impl SealedSessionCipher {
+    pub(crate) fn new(
+        context: Context,
+        store_context: StoreContext,
+        local_address: ServiceAddress,
+        certificate_validator: CertificateValidator,
+    ) -> Self {
+        Self {
+            context,
+            store_context,
+            local_address,
+            certificate_validator,
+        }
+    }
+
+    #[allow(dead_code)]
```

Comment:
Why is that? Maybe add a comment explaining, if still needed.

Reply:
Oooh, because I haven't implemented _sending_ of encrypted messages in the `MessageSender` yet.

Reply:
Should that still go in this PR, or should that be an issue? If the latter, I suggest writing a comment on the `allow(dead_code)` linking to the issue.